pennylane/transforms/decompositions/clifford_plus_t/solovay_kitaev.py

```diff
@@ -0,0 +1,377 @@
+
+        return dist > tol
+
+
+def _unitary_bloch(mat, eps=1e-10):
+    """Computes angle and axis for any given unitary matrix."""
+
+    angle = qml.math.real(qml.math.arccos(qml.math.trace(mat) / 2))
+    sine = qml.math.sin(angle)
+
+    if sine < eps:
+        axis = [0, 0, 1]
+    else:
+        nx = -qml.math.imag((mat[1, 0] + mat[0, 1]) / (2 * sine))
+        ny = qml.math.real((mat[1, 0] - mat[0, 1]) / (2 * sine))
+        nz = qml.math.imag((mat[1, 1] - mat[0, 0]) / (2 * sine))
+        axis = [nx, ny, nz]
+
+    return axis, 2 * angle
+
+
+def _group_commutator_decompose(mat):
+    """Performs group commutator decomposition U = V' @ W' @ V'.dag @ W'.dag."""
+    # Get axis and theta for the operator.
+    axis, theta = _unitary_bloch(mat)
+
+    # The angle phi comes from eq 10 in 'The Solovay-Kitaev Algorithm' paper
+    phi = 2.0 * qml.math.arcsin(qml.math.sqrt(qml.math.sqrt((0.5 - 0.5 * qml.math.cos(theta / 2)))))
+
+    # Begin decomposition by computing the rotation matrices
+    v = qml.RX(phi, [0])
+    w = qml.RY(2 * math.pi - phi, [0]) if axis[2] > 0 else qml.RY(phi, [0])
+
+    # Get similarity transormation matrix S and S.dag
+    ud = qml.math.linalg.eig(mat)[1]
+    vwd = qml.math.linalg.eig(qml.matrix(v @ w @ v.adjoint() @ w.adjoint()))[1]
+    s = ud @ qml.math.conj(qml.math.transpose(vwd))
+    sdg = vwd @ qml.math.conj(qml.math.transpose(ud))
+
+    # Get the required matrices V' and W'
+    v_hat = s @ v.matrix() @ sdg
+    w_hat = s @ w.matrix() @ sdg
+
+    return w_hat, v_hat
+
+
+def _approximate_umat(seqs, basic_approximations):
+    """Approximates given GateSet using the TreeSet structure"""
+
+    def key(x):
+        return qml.math.linalg.norm(qml.math.subtract(x.so3_matrix, seqs.so3_matrix))
+
+    return min(basic_approximations, key=key)
+
+
+def sk_approximate_set(basis_set=None, depth=10):
+    r"""Builds approximate unitary set required for Solovay-Kitaev algorithm.
+
+    Args:
+        basis_set (list(str)): Basis set to be used for Solovay-Kitaev decomposition build using
+            following terms, ``['x', 'y', 'z', 'h', 't', 'tdg', 's', 'sdg']``
+        depth (int): Maximum expansion length of Clifford+T sequences in the approximation set
+
+    Returns:
+        list: A list of Clifford+T sequences that will be used for approximating a matrix in the base case of recursive implementation of Solovay-Kitaev.
+    """
+
+    if basis_set is None:
+        basis_set = ["t", "tdg", "h"]
+
+    return TreeSet(GateSet(), basis_set, ()).basic_approximation(depth)
+
+
+def sk_decomposition(op, depth, basis_set=None, approximate_set=None):
+    r"""Approximate an arbitrary single-qubit gate in Clifford+T basis using `Solovay-Kitaev algorithm <https://arxiv.org/abs/quant-ph/0505030>`_.
+
+    This method implements a recursive Solovay-Kitaev decomposition for approximating any SU(2) operation
+    with :math:`\epsilon > 0` error that depdends on the recursion ``depth``. In general, this algorithm runs
```

Comment:
> error that depends on the recursion ``depth``

I think readers will probably ask "how can I know the depth for my required error?"
Do we know this? Would anyone want to input the error instead of the depth?

Reply:
I don't think it is that straightforward to calculate that. Depth would depend on the approximation set we use in the base case. If that is sufficiently big, one should be fine with even low recursion depths.